### PR TITLE
Add topic loading throttling

### PR DESCRIPTION
**Pull request by rdhabalia**
### Motivation

Addressing #300 : At the time of broker cold-restart, clients make broker to load all the topics concurrently which can create back pressure at zk and may increase zk latency. So, broker should have a way to throttle number of concurrent topic loading.

### Modifications

- Introduce dynamic throttling limit to control concurrent topic loading.
- above that limit broker rejects producer/consumer create request so, producer and consumer can do retry with backoff

### Result

It will help broker to control number of zk operations on cold-restart.

### CODE REVIEW

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java [hunk 1]
```diff
@@ -458,20 +467,23 @@ public void openLedgerComplete(ManagedLedger ledger, Object ctx) {
 
                                 return null;
                             });
+                            topicLoadRequestSemaphore.get().release();
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: The `acquire()` and `release()` operations might not happen on the same semaphore if the limit is being changed.

You should save the instance into a local variable and call release on the same object.

**Comment by rdhabalia** [comment]
> If we reject the loading, an unused topic will not get loaded later. That's why I was also thinking of "queuing" the topic load operations.

That's correct. However, if topic is unused (no producer/consumer) then isn't it deleted by GC from that broker. So, after that it will not reflect into stats. And if there is any producer wants to connect then anyway that topic will be loaded after sometime. 
Is that correct?

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java (hunk 1 above)
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: yes, that make sense. updated it.

**Comment by merlimat** [comment]
> However, if topic is unused (no producer/consumer) then isn't it deleted by GC from that broker. 

The topic can have subscriptions and no producers/consumers using it

**Comment by rdhabalia** [comment]
> The topic can have subscriptions and no producers/consumers using it

That's true, GC doesn't delete topic with non-empty subscription so, we can get the stat. Updated the change.

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java [hunk 2]
```diff
@@ -430,6 +452,18 @@ public PulsarClient getReplicationClient(String cluster) {
         }
 
         final CompletableFuture<Topic> topicFuture = new CompletableFuture<>();
+        final Semaphore topicLoadSemaphore = topicLoadRequestSemaphore.get();
+
+        if (!topicLoadSemaphore.tryAcquire()) {
+            if (failOnTooManyConcurrentTopics) {
+                throw new TooManyRequestsException("Too many concurrent topic load exception");
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: Couldn't we just return a failed future at this point?

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java [hunk 3]
```diff
@@ -418,6 +431,15 @@ public PulsarClient getReplicationClient(String cluster) {
     }
 
     private CompletableFuture<Topic> createPersistentTopic(final String topic) throws RuntimeException {
+        try {
+            return createPersistentTopic(topic, false);
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: It is not immediate to understand what `false` means here

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java [hunk 4]
```diff
@@ -870,7 +913,10 @@ private void updateConfigurationAndRegisterListeners() {
         updateDynamicServiceConfiguration();
         // add listener on "maxConcurrentLookupRequest" value change
         registerConfigurationListener("maxConcurrentLookupRequest",
-                (pendingLookupRequest) -> lookupRequestSemaphore.set(new Semaphore((int) pendingLookupRequest, true)));
+                (maxConcurrentLookupRequest) -> lookupRequestSemaphore.set(new Semaphore((int) maxConcurrentLookupRequest, true)));
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: I think we shouldn't strictly require the "fairness" in the semaphore.

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java [hunk 5]
```diff
@@ -972,4 +1018,31 @@ public void onUpdate(String path, Map<String, String> data, Stat stat) {
         return dynamicConfigurationMap;
     }
 
-}
+    private void processPendingLoadTopic() {
+        lock.lock();
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: Why do we need this lock? Would it be enough to `poll()` one item out of the queue and process it?

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java [hunk 6]
```diff
@@ -972,4 +1018,31 @@ public void onUpdate(String path, Map<String, String> data, Stat stat) {
...
+        try {
+            Pair<String, CompletableFuture<Topic>> pendingTopic = pendingTopicLoadingQueue.poll();
+            if (pendingTopic != null) {
+                try {
+                    createPersistentTopic(pendingTopic.getLeft(), true).handle((topic, ex) -> {
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: We should check if we still own the topic

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java [hunk 7]
```diff
@@ -418,6 +431,15 @@ public PulsarClient getReplicationClient(String cluster) {
...
 
     private CompletableFuture<Topic> createPersistentTopic(final String topic) throws RuntimeException {
+        try {
+            return createPersistentTopic(topic, false);
+        } catch (TooManyRequestsException e) {
+            return FutureUtil.failedFuture(e);
+        }
+    }
+
+    private CompletableFuture<Topic> createPersistentTopic(final String topic, boolean failOnTooManyConcurrentTopics)
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: Instead of having these 2 methods, I was thinking something like:

```java
CompletableFuture<Topic> createPersistentTopic(final String topic) {
    // Creates future and call method below or queue the request
}


void createPersistentTopic(String topic, CompletableFuture<Topic> topicFuture) {
   // To be called when we are ready to load the topic
} 
```

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java [hunk 8]
```diff
@@ -972,4 +1018,31 @@ public void onUpdate(String path, Map<String, String> data, Stat stat) {
...
+                try {
+                    createPersistentTopic(pendingTopic.getLeft(), true).handle((topic, ex) -> {
+                        if (ex != null) {
+                            pendingTopic.getRight().completeExceptionally(ex);
+                        } else {
+                            pendingTopic.getRight().complete(topic);
+                        }
+                        return null;
+                    });
+                } catch (TooManyRequestsException te) {
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: We shouldn't even have to check again at this point.

Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: what if this time(as mentioned in above comment) some other new topic3 (from different bundle) comes and acquired a permit and there could be possibility that creation of this topic may fail again with `TooManyRequestsException`. So, we have to again put it back to the queue, as we don't want to fail it and try again.

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java (hunk 5 above)
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: it is to prevent race condition which may cause topic reside in `pendingTopicLoadingQueue` forever.
eg:
- if `MaxConcurrentTopicLoadRequest`=1 and if two topics (topic1, topic2) come at same time then one of them (topic2) will be added into pendingQueue. 
- as soon as topic1 completes: it tries to release semaphore and `poll` topic2 from queue to createTopic again, but if 3rd new topic3 acquired a permit before it and when it releases it may find `pendingTopicLoadingQueue` empty as we have already `poll` topic2 it before. after that topic2 will be added again on [line-1037](https://github.com/rdhabalia/pulsar/blob/406b0fc6b8efcc1451e163858a1e3f72860968fd/pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java#L1037).

So, added lock as [pendingTopicLoadingQueue.poll()](https://github.com/rdhabalia/pulsar/blob/406b0fc6b8efcc1451e163858a1e3f72860968fd/pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java#L1024) and [pendingTopicLoadingQueue.add(pendingTopic)](https://github.com/rdhabalia/pulsar/blob/406b0fc6b8efcc1451e163858a1e3f72860968fd/pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java#L1037) is not atomic.
Do you think we should ignore this situation?

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java (hunk 2 above)
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: the main idea behind failing immediately with exception so, while retrying create topic from [pendingTopicLoadingQueue](https://github.com/rdhabalia/pulsar/blob/406b0fc6b8efcc1451e163858a1e3f72860968fd/pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java#L1037) if it fails again here then we can put same topic-future again into the queue because it has original `future` which we had returned back to `ServerCnx` on very first time when `createPersistentTopic` request came.

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java (hunk 3 above)
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: yes, we can add comment. But the idea of having this flag:
1. when `BrokerService` receives `createPersistentTopic` request, passing `failOnTooManyConcurrentTopics=false` means we will not fail the request due to throttling but will add into the queue.
2. But when we retry from `pendingTopicLoadingQueue` and again if it gets throttle then we add that topic-future again into the queue at [line1037](https://github.com/rdhabalia/pulsar/blob/406b0fc6b8efcc1451e163858a1e3f72860968fd/pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java#L1037)

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java (hunk 4 above)
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: you means we should not create a new semaphore? I have set the new instance rather updating permit into existing semaphore so, we don't mess up with new permits (reduce/increment) number as `lookupRequestSemaphore ` is already concurrently used by multiple threads.

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java (hunk 7 above)
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]:
```
void createPersistentTopic(String topic, CompletableFuture<Topic> topicFuture) {
   // To be called when we are ready to load the topic
}
```
Does this method throttle again and puts in the queue again or it just fails the future if semaphore doesn't have any additional permits?

File: pulsar-broker/src/test/java/com/yahoo/pulsar/client/impl/BrokerClientIntegrationTest.java [hunk 9]
```diff
@@ -517,6 +523,65 @@ public void testCloseConnectionOnBrokerRejectedRequest() throws Exception {
...
+                futures.add(pulsarClient2.createProducerAsync(topicName + randomUUID().toString()));
+                futures.add(pulsarClient.createProducerAsync(topicName + randomUUID().toString()));
+                latch.countDown();
+            });
+        }
+
+        latch.await();
+        FutureUtil.waitForAll(futures).get();
+
+        pulsarClient.close();
```
Comment (saandrews) [inline pulsar-broker/src/test/java/com/yahoo/pulsar/client/impl/BrokerClientIntegrationTest.java]: Can you add test to verify negative scenario with TooManyReqeusts exception?

Comment (rdhabalia) [inline pulsar-broker/src/test/java/com/yahoo/pulsar/client/impl/BrokerClientIntegrationTest.java]: actually as we decided broker should not give `TooManyReqeustsException` to client and put the topic in to the queue so, it give chance to unused topic to be loaded for stats visibility. So, eventually all the producers should get connected even it gets error internally at broker.

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java (hunk 7 above)
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: @merlimat I think I updated changes as per your suggestion

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java [hunk 10]
```diff
@@ -417,20 +426,55 @@ public PulsarClient getReplicationClient(String cluster) {
...
+     */
     private CompletableFuture<Topic> createPersistentTopic(final String topic) throws RuntimeException {
         checkTopicNsOwnership(topic);
 
+        final CompletableFuture<Topic> topicFuture = new CompletableFuture<>();
+
+        final Semaphore topicLoadSemaphore = topicLoadRequestSemaphore.get();
+
+        if (topicLoadSemaphore.tryAcquire()) {
+            CompletableFuture<Topic> future = new CompletableFuture<>();
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: I think this could be done with just 1 future, just the same way you add the handle method to attach behavior after completing

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java (hunk 4 above)
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: No no, I was referring to the "true" as 2nd parameter for the semaphore constructor. That's about the fairness of the semaphore itself. I think we might not need the exact fairness, so we could relax that option.

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java [hunk 11]
```diff
@@ -972,4 +1017,41 @@ public void onUpdate(String path, Map<String, String> data, Stat stat) {
...
-}
+    /**
+     * Create pending topic and on completion it picks the next one until processes all topics in
+     * {@link #pendingTopicLoadingQueue}.<br/>
+     * It also tries to acquire {@link #topicLoadRequestSemaphore} so throttle down newly incoming topics and release
+     * permit if it was successful to acquire it.
+     */
+    private void createPendingLoadTopic() {
+        Pair<String, CompletableFuture<Topic>> pendingTopic = pendingTopicLoadingQueue.poll();
+        if (pendingTopic != null) {
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: if (pendingtopic == null) return;

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java [hunk 12]
```diff
@@ -972,4 +1017,41 @@ public void onUpdate(String path, Map<String, String> data, Stat stat) {
...
+            final String topic = pendingTopic.getLeft();
+            try {
+                checkTopicNsOwnership(topic);
+                CompletableFuture<Topic> future = new CompletableFuture<>();
+                final Semaphore topicLoadSemaphore = topicLoadRequestSemaphore.get();
+                final boolean acquiredPermit = topicLoadSemaphore.tryAcquire();
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: What is the reason for acquiring the semaphore here, if we're not waiting for it anyway?

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java [hunk 13]
```diff
@@ -972,4 +1017,41 @@ public void onUpdate(String path, Map<String, String> data, Stat stat) {
...
+     * It also tries to acquire {@link #topicLoadRequestSemaphore} so throttle down newly incoming topics and release
+     * permit if it was successful to acquire it.
+     */
+    private void createPendingLoadTopic() {
+        Pair<String, CompletableFuture<Topic>> pendingTopic = pendingTopicLoadingQueue.poll();
+        if (pendingTopic != null) {
+            final String topic = pendingTopic.getLeft();
+            try {
+                checkTopicNsOwnership(topic);
+                CompletableFuture<Topic> future = new CompletableFuture<>();
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: Same as above, problably we don't need an additional future

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java [hunk 14]
```diff
@@ -972,4 +1017,41 @@ public void onUpdate(String path, Map<String, String> data, Stat stat) {
...
+                        topicLoadSemaphore.release();
+                    }
+                    createPendingLoadTopic();
+                    return null;
+                });
+            } catch (RuntimeException re) {
+                log.error("Failed to create pending topic {} {}", topic, re);
+                pendingTopic.getRight().completeExceptionally(re.getCause());
+                // process next pending topic
+                createPendingLoadTopic();
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: This would be a recursive call

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java (hunk 12 above)
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: actually with this approach compare to my previous [commit-406b0fc](https://github.com/yahoo/pulsar/pull/306/commits/406b0fc6b8efcc1451e163858a1e3f72860968fd) fails to provide throttling in certain usecases.
eg:
1. maxConcurrentTopicLoadRequest=10 and broker receives 21 requests so, 10 will be executed and rest of them will go to pending-queue
2. once first 10 completes it will try to execute pending-10 without any throttling. but at the same time if additional 31 requests come then 10 requests out of them will be served as semaphore is already released. so, now we are serving total 20 concurrent request (10(new requests) + 10(10 completed topic threads which invoked `createPendingLoadTopic`)
3. and this process go on which can increase concurrent-count and we are not doing real throttling.

- However, I am acquiring semaphore here so, at least those new additional coming 31 requests will be throttled. Still this approach doesn't throttle 100% in compare to [commit-406b0fc](https://github.com/yahoo/pulsar/pull/306/commits/406b0fc6b8efcc1451e163858a1e3f72860968fd)

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java (hunk 14 above)
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: `createPersistent` is async call so, we are making it recursive once future is completed otherwise due to throttling-limit=10 if queue has 1000 pending requests then all 1000 can go concurrently through `createPersistent(topic,future)` . So, we have to call it recursively only after receiving completed future. right?

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java (hunk 12 above)
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: Ok, that makes sense

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java (hunk 14 above)
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: But in this case if you get an exception when *starting* the create topic asynchronous, t will immediately start the next one. If all th queued topics start failing we'd likely get into stack overflow errors. Easy fix would be to schedule the next topic load from executor

Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: that make sense.

**Comment by merlimat** [comment]
Couple of seemingly legitimate unit test failures: 

```
ServerCnxTest.testProducerOnNotOwnedTopic:397 expected [ServiceNotReady] but found [UnknownError]
  BrokerClientIntegrationTest.testMaxConcurrentTopicLoading:579 » Execution com....
```

**Comment by rdhabalia** [comment]
yes, fixed.
